iceoryx_posh/source/roudi/roudi_process.cpp

```diff
@@ -126,21 +134,154 @@ ProcessManager::ProcessManager(RouDiMemoryInterface& roudiMemoryInterface,
+        processStillRunning[i] = requestShutdownOfProcess(process, ShutdownPolicy::SIG_TERM, ShudownLog::FULL);
+        ++i;
+    }
 
-    for (; itEnd != it; ++it)
+    // wait for process to complete
+    bool allProcessFinished = false;
+    posix::Timer finalKillTimer(processKillDelay);
+
+    yield();
+    auto waitForKillProcessed = [&]() {
+        bool processStateChanged = true;
+        finalKillTimer.resetCreationTime();
+        while (!allProcessFinished && !finalKillTimer.hasExpiredComparedToCreationTime())
+        {
+            i = 0;
+            for (auto& process : m_processList)
+            {
+                if (processStillRunning[i] && isProcessTerminated(process))
+                {
+                    processStillRunning[i] = false;
+                    processStateChanged = true;
+                }
+                ++i;
+            }
+
+            // check if we are done
+            if (processStateChanged)
+            {
+                processStateChanged = false;
+                allProcessFinished = true;
+                for (bool isRunning : processStillRunning)
+                {
+                    allProcessFinished &= !isRunning;
+                }
+            }
+
+            if (!allProcessFinished)
+            {
+                yield();
+            }
+        }
+    };
+    waitForKillProcessed();
+
+    // any processes still alive? Time to send SIG_KILL to kill.
+    if (finalKillTimer.hasExpiredComparedToCreationTime())
     {
-        if (-1 == kill(static_cast<pid_t>(it->getPid()), SIGTERM))
+        i = 0;
+        for (auto& process : m_processList)
         {
-            LogWarn() << "Process " << it->getPid() << " could not be killed";
+            if (processStillRunning[i])
+            {
+                LogWarn() << "Process ID " << process.getPid() << " named '" << process.getName()
+                          << "' is still running after SIGTERM was sent " << processKillDelay.seconds<int>()
+                          << " seconds ago. RouDi is sending SIGKILL now.";
+                processStillRunning[i] = requestShutdownOfProcess(process, ShutdownPolicy::SIG_KILL, ShudownLog::FULL);
+            }
+            ++i;
         }
+
+        // we sent SIG_KILL to kill, now wait till they have terminated
+        waitForKillProcessed();
+        if (finalKillTimer.hasExpiredComparedToCreationTime())
+        {
+            i = 0;
+            for (auto& process : m_processList)
+            {
+                if (processStillRunning[i])
+                {
+                    LogWarn() << "Process ID " << process.getPid() << " named '" << process.getName()
+                              << "' is still running after SIGKILL was sent " << processKillDelay.seconds<int>()
+                              << " seconds ago. RouDi is ignoring this process.";
+                }
+                ++i;
+            }
+        }
+    }
+
+    auto it = m_processList.begin();
```

Comment:
What happens when `m_processList` is empty. Then `it` would point to `m_processList.end()` since it is equal to `begin` in this case and `removeProcess` would segfault since we are accessing an invalid element.

Reply:
the method checks if `it` is `iterator.end()` so nothing bad happens. But you are right, it's not obvious from the function call